[SPARK-37655][CORE] Add `RocksDB` Implementation for `KVStore`

dongjoon-hyun opened the pull request:
### What changes were proposed in this pull request?

This PR aims to add `RocksDB` implementation for `KVStore` which is a clone of `LevelDB` implementation.
In the future, we can have a `LevelDB`-free Apache Spark distribution. 
Please note that this PR intentionally add `RocksDB`-only code and avoid LevelDB-related code changes (like refactoring).

### Why are the changes needed?

Since Apache Spark 2.3.0 (SPARK-20641) in 2017, we have `LevelDB` implementation for `KVStore` interface.
Currently, `LevelDB` JNI library became inactive and blocks Apple Silicon support.
Instead, `RocksDB` community is actively working on Apple Silicon at https://github.com/facebook/rocksdb/issues/7720 .

### Does this PR introduce _any_ user-facing change?

No. This is used in Spark internal.

### How was this patch tested?

Pass the CIs.

The benchmark seems to be not supported yet in our GitHub Action Benchmark job.
I attached the performance comparison on `MacBook Pro (15-inch 2019, 2.6 GHz 6-Core Intel Core i7)`.

```
ProductName:	macOS
ProductVersion:	12.0.1
BuildVersion:	21A559
```

```
[INFO] Running org.apache.spark.util.kvstore.LevelDBBenchmark
                                        	count	mean	min	max	95th
dbClose                                 	4	1.030	0.963	1.124	1.124
dbCreation                              	4	90.072	3.352	354.065	354.065
naturalIndexCreateIterator              	1024	0.009	0.003	2.289	0.011
naturalIndexDescendingCreateIterator    	1024	0.005	0.004	0.043	0.007
naturalIndexDescendingIteration         	1024	0.011	0.009	0.037	0.013
naturalIndexIteration                   	1024	0.012	0.009	0.099	0.015
randomDeleteIndexed                     	1024	0.033	0.028	0.190	0.039
randomDeletesNoIndex                    	1024	0.017	0.013	0.036	0.018
randomUpdatesIndexed                    	1024	0.124	0.054	37.057	0.122
randomUpdatesNoIndex                    	1024	0.051	0.034	0.649	0.097
randomWritesIndexed                     	1024	0.187	0.073	64.647	0.210
randomWritesNoIndex                     	1024	0.068	0.040	1.567	0.128
refIndexCreateIterator                  	1024	0.005	0.004	0.019	0.007
refIndexDescendingCreateIterator        	1024	0.003	0.003	0.018	0.005
refIndexDescendingIteration             	1024	0.012	0.009	0.052	0.014
refIndexIteration                       	1024	0.015	0.010	0.362	0.016
sequentialDeleteIndexed                 	1024	0.031	0.026	0.123	0.035
sequentialDeleteNoIndex                 	1024	0.017	0.013	0.114	0.019
sequentialUpdatesIndexed                	1024	0.049	0.044	0.777	0.054
sequentialUpdatesNoIndex                	1024	0.076	0.038	0.793	0.113
sequentialWritesIndexed                 	1024	0.061	0.049	1.885	0.104
sequentialWritesNoIndex                 	1024	0.063	0.044	2.411	0.120
```

```
[INFO] Running org.apache.spark.util.kvstore.RocksDBBenchmark
                                        	count	mean	min	max	95th
dbClose                                 	4	1.842	0.611	5.300	5.300
dbCreation                              	4	142.354	5.993	565.837	565.837
naturalIndexCreateIterator              	1024	0.009	0.004	2.109	0.012
naturalIndexDescendingCreateIterator    	1024	0.008	0.006	0.086	0.011
naturalIndexDescendingIteration         	1024	0.013	0.008	0.073	0.020
naturalIndexIteration                   	1024	0.013	0.009	0.195	0.019
randomDeleteIndexed                     	1024	0.242	0.133	0.647	0.344
randomDeletesNoIndex                    	1024	0.188	0.101	0.437	0.284
randomUpdatesIndexed                    	1024	0.340	0.196	37.788	0.406
randomUpdatesNoIndex                    	1024	0.268	0.151	4.927	0.392
randomWritesIndexed                     	1024	0.431	0.210	75.838	0.467
randomWritesNoIndex                     	1024	0.248	0.150	2.033	0.359
refIndexCreateIterator                  	1024	0.008	0.005	0.059	0.012
refIndexDescendingCreateIterator        	1024	0.004	0.004	0.025	0.006
refIndexDescendingIteration             	1024	0.013	0.010	0.057	0.016
refIndexIteration                       	1024	0.017	0.011	0.376	0.026
sequentialDeleteIndexed                 	1024	0.203	0.131	0.543	0.272
sequentialDeleteNoIndex                 	1024	0.212	0.122	3.392	0.314
sequentialUpdatesIndexed                	1024	0.265	0.129	0.994	0.388
sequentialUpdatesNoIndex                	1024	0.252	0.155	0.963	0.365
sequentialWritesIndexed                 	1024	0.267	0.155	2.382	0.372
sequentialWritesNoIndex                 	1024	0.253	0.137	2.695	0.371
```

dongjoon-hyun committed:
[SPARK-37655][CORE] Add RocksDB Implementation for KVStore

SparkQA commented:
**[Test build #146245 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146245/testReport)** for PR 34913 at commit [`4b603fd`](https://github.com/apache/spark/commit/4b603fd34c571d65da3e060d617b4694707b7b29).

SparkQA commented:
**[Test build #146245 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146245/testReport)** for PR 34913 at commit [`4b603fd`](https://github.com/apache/spark/commit/4b603fd34c571d65da3e060d617b4694707b7b29).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class RocksDB implements KVStore `
  * `  public static class TypeAliases `
  * `class RocksDBIterator<T> implements KVStoreIterator<T> `
  * `class RocksDBTypeInfo `
  * `  class Index `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146245/

dongjoon-hyun committed:
fix Java linter

dongjoon-hyun commented:
cc @viirya , @HyukjinKwon , @LuciferYang

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50719/

SparkQA commented:
**[Test build #146246 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146246/testReport)** for PR 34913 at commit [`dd3d009`](https://github.com/apache/spark/commit/dd3d0099ac57654ef22f834b26ad34e8480adabf).

viirya commented:
Thanks. I will take a look.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50719/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50720/

dongjoon-hyun commented:
Thank you, @viirya .

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50719/

dongjoon-hyun commented:
BTW, Scala 2.13 failure is irrelevant to this PR. The root cause is reverted from master branch 25 minutes ago.
- https://github.com/apache/spark/commit/16841286ecb31446300c81e3f98516e92846d9fa

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50720/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50720/

SparkQA commented:
**[Test build #146246 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146246/testReport)** for PR 34913 at commit [`dd3d009`](https://github.com/apache/spark/commit/dd3d0099ac57654ef22f834b26ad34e8480adabf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146246/

LuciferYang commented:
Very valuable changes, I will take a look

LuciferYang commented on common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDB.java:
```diff
@@ -0,0 +1,436 @@
+  /** DB key where type aliases are stored. */
+  private static final byte[] TYPE_ALIASES_KEY = "__types__".getBytes(UTF_8);
+
+  private static final BlockBasedTableConfig tableFormatConfig = new BlockBasedTableConfig()
```
Does `tableFormatConfig` and `dbOptions` need to be class variables? It seems that they will only be used when initializing `_db`

dongjoon-hyun replied:
Thank you for review, @LuciferYang .
Actually, `BlockBasedTableConfig` has a set of configurations and I tested some of them and cleaned up now.
When we tune it in the future, this style is more easier to change `BlockBasedTableConfig` only instead of touching `dbOptions`.

LuciferYang replied:
Thank you for your explanation.

LuciferYang commented on common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDB.java:
```diff
@@ -0,0 +1,436 @@
+
+  private static final WriteOptions writeOptions = new WriteOptions().setSync(true);
+
+  private AtomicReference<org.rocksdb.RocksDB> _db;
```
Should `_db` be identified as `final`? It is assigned only once

LuciferYang commented on common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDB.java:
```diff
@@ -0,0 +1,436 @@
+
+  }
+
+  private static class PrefixCache {
```
`PrefixCache` is repeatedly defined here because this pr does not want to introduce too many refactoring work, right?

dongjoon-hyun replied:
Yes, exactly, @LuciferYang . You can do `diff LevelDB.java RocksDB.java` to see the difference.
We are going to remove LevelDB in the future.

LuciferYang replied:
OK ~

LuciferYang commented on common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDBIterator.java:
```diff
@@ -0,0 +1,277 @@
+import com.google.common.base.Throwables;
+import org.rocksdb.RocksIterator;
+
+import java.io.IOException;
```
This import order needs fix

dongjoon-hyun replied:
Ya, right. Java `import` should be at the first. Thanks!

LuciferYang commented on common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDBIterator.java:
```diff
@@ -0,0 +1,277 @@
+      if (it.isValid() && compare(it.key(), indexKeyPrefix) <= 0) {
+        // continue
+      } else {
+        it.prev();
```
For `LevelDBIterator`, there will call `it.peekNext().getKey()` to get `nextKey`, and `peekNext` method will throw `NoSuchElementException` if `iterator.isValid()` is false.

For `RocksIterator`, if `it.isValid()` if false will call `it.prev()`,  will this operation throw other exceptions?

LuciferYang replied:
so should we change to 

```
if(!it.isValid()) {
        throw new NoSuchElementException();
      }
      if (compare(it.key(), indexKeyPrefix) > 0) {
        it.prev();
      }
```

LuciferYang commented on common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDBIterator.java:
```diff
@@ -0,0 +1,277 @@
+   * when Scala wrappers are used, this makes sure that, hopefully, the JNI resources held by
+   * the iterator will eventually be released.
+   */
+  @SuppressWarnings("deprecation")
```
nit: we can remove this redundant suppression

LuciferYang commented on common/kvstore/src/test/java/org/apache/spark/util/kvstore/RocksDBBenchmark.java:
```diff
@@ -0,0 +1,280 @@
+      System.out.println(row);
+    }
+
+    Slf4jReporter.forRegistry(metrics).outputTo(LoggerFactory.getLogger(LevelDBBenchmark.class))
```
LevelDBBenchmark.class -> RocksDBBenchmark.class

dongjoon-hyun committed:
Fix import order

dongjoon-hyun committed:
Address comments

dongjoon-hyun committed:
Add final

dongjoon-hyun commented:
Thank you, @LuciferYang . I addressed your comments.

dongjoon-hyun commented:
Could you review this once more, @viirya and @LuciferYang ?

SparkQA commented:
**[Test build #146309 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146309/testReport)** for PR 34913 at commit [`405b84c`](https://github.com/apache/spark/commit/405b84c66ee5bc7154b2ff2583b0b9e90dc00649).

LuciferYang reviewed (approved):
LGTM

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50781/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50781/

BelodengKlaus commented:
the performance shows that LevelDB is faster than RocksDB. And after Apple Silicon support RocksDB may better than LevelDB ?
(maybe only reason is the LevelDB became inactive?)

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50781/

SparkQA commented:
**[Test build #146309 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146309/testReport)** for PR 34913 at commit [`405b84c`](https://github.com/apache/spark/commit/405b84c66ee5bc7154b2ff2583b0b9e90dc00649).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146309/

dongjoon-hyun commented:
@BelodengKlaus 
> the performance shows that LevelDB is faster than RocksDB. And after Apple Silicon support RocksDB may better than LevelDB ? (maybe only reason is the LevelDB became inactive?)

Yes, currently, both DB doesn't work at all at native Java distribution on Apple Silicon.

At Apache Spark 3.3, we will have both backends and the user can choose one of them for their use cases. The performance tuning will be followed too.

gengliangwang commented:
@dongjoon-hyun just curious about the whole plan: what will happen if a user enables the RocksDB KVStore when there are existing LevelDB files?

BelodengKlaus commented:
> @BelodengKlaus
> 
> > the performance shows that LevelDB is faster than RocksDB. And after Apple Silicon support RocksDB may better than LevelDB ? (maybe only reason is the LevelDB became inactive?)
> 
> Yes, currently, both DB doesn't work at all at native Java distribution on Apple Silicon.
> 
> At Apache Spark 3.3, we will have both backends and the user can choose one of them for their use cases. The performance tuning will be followed too.

LGTM

dongjoon-hyun commented:
I don't think we need to support that kind of migration, @gengliangwang .
> what will happen if a user enables the RocksDB KVStore when there are existing LevelDB files?

The main goal in Apache Spark 3.3 is to support new installation on M1-based machine. On M1-based machine, both LevelDB and RocksDB do not work. So, there is no previous data on that machines.

LuciferYang commented:
> @dongjoon-hyun just curious about the whole plan: what will happen if a user enables the RocksDB KVStore when there are existing LevelDB files?

As a user who uses spark historyserver in a relatively large range (about 150000+ SparkApp every day), I accept deleting the `LevelDB` instances and reply them again during the historyserver upgrade process

BelodengKlaus commented on common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDB.java:
```diff
@@ -0,0 +1,436 @@
+        try (WriteBatch writeBatch = new WriteBatch()) {
+          while (valueIter.hasNext()) {
+            updateBatch(writeBatch, valueIter.next(), serializedValueIter.next(), klass,
+                    naturalIndex, indices);
```
indent

klaus-xiong reviewed (commented):
why we dont have a RockDBTypeInfoSuite like LevelDB?

viirya reviewed (commented):
I'm not familiar with the related code of KVStore. Looks like this is basically following the LevelDB's KVStore implementation, but only with RocksDB as underlying database store.

So what is the plan for two KVStores? Will we keep both or remove LevelDB KVStore in the next step?

dongjoon-hyun committed:
Add RocksDBTypeInfoSuite

dongjoon-hyun committed:
indextation

SparkQA commented:
**[Test build #146338 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146338/testReport)** for PR 34913 at commit [`abb6b6d`](https://github.com/apache/spark/commit/abb6b6dc102ffd4a214b9a016f21edfb7929f901).

dongjoon-hyun commented:
To @LuciferYang , it's possible, but it's beyond of the scope of this PR.
To @klaus-xiong , thank you for review. I added `RocksDBTypeInfoSuite` and addressed the indentation part too.
To @viirya , yes. 
- We will keep both in Apache Spark 3.3 and the backend will be configurable like `spark.sql.orc.impl`.
- I'm working on removing them with `ExtendedRocksDBTest` together in Apache Spark 3.3 timeframe.

viirya reviewed (commented):
This should be good as it simply uses RocksDB to replace LevelDB's role in its KVStore implementation. Otherwise the implementation details look the same with LevelDB's KVStore.

I checked the current status of LevelDB JNI. It is inactive as the description mentioned. Although I also saw some users asking about Apple Silicon support for it, but I don't see any response after months.

Due to the situation, RocksDB seems to be good replacement here.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50812/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50812/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50812/

dongjoon-hyun commented:
Thank you all. All tests passed although the UI doesn't show it yet. Merged to master.

<img width="748" alt="Screen Shot 2021-12-17 at 11 37 26 AM" src="https://user-images.githubusercontent.com/9700541/146598745-4f882877-aece-4ff8-94aa-5eacb716758c.png">
.

SparkQA commented:
**[Test build #146338 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146338/testReport)** for PR 34913 at commit [`abb6b6d`](https://github.com/apache/spark/commit/abb6b6dc102ffd4a214b9a016f21edfb7929f901).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146338/

mridulm commented:
+CC @thejdeep, @shardulm94 FYI
